lib/src/barback/dartdevc/dartdevc_environment.dart

```diff
@@ -0,0 +1,266 @@
+      }
+    } else if (id.path.endsWith('require.js') ||
+        id.path.endsWith('dart_sdk.js')) {
+      assets = {id: _buildJsResource(id)};
+    } else if (id.path.endsWith('require.js.map') ||
+        id.path.endsWith('dart_sdk.js.map')) {
+      assets = {id: new Future.error(new AssetNotFoundException(id))};
+    } else if (id.path.endsWith('.js') || id.path.endsWith('.js.map')) {
+      var jsId = id.extension == '.map' ? id.changeExtension('') : id;
+      assets = createDartdevcModule(jsId, _moduleReader, _scratchSpace,
+          _environmentConstants, _mode, logError);
+    } else if (id.path.endsWith(moduleConfigName)) {
+      assets = {id: _buildModuleConfig(id)};
+    }
+    assets ??= <AssetId, Future<Asset>>{};
+
+    for (var id in assets.keys) {
+      _assetCache[id] = assets[id];
+    }
+    return assets;
+  }
+
+  /// Builds a module config asset at [id].
+  Future<Asset> _buildModuleConfig(AssetId id) async {
+    assert(id.path.endsWith(moduleConfigName));
+    var moduleDir = topLevelDir(id.path);
+    var allAssets = await _barback.getAllAssets();
+    var moduleAssets = allAssets.where((asset) =>
+        asset.id.package == id.package &&
+        asset.id.extension == '.dart' &&
+        topLevelDir(asset.id.path) == moduleDir);
+    var moduleMode =
+        moduleDir == 'lib' ? ModuleMode.public : ModuleMode.private;
+    var modules = await computeModules(moduleMode, moduleAssets);
+    var encoded = JSON.encode(modules);
+    return new Asset.fromString(id, encoded);
+  }
+
+  /// Builds the `dart_sdk.js` or `require.js` assets by copying them from the
+  /// sdk.
+  Future<Asset> _buildJsResource(AssetId id) async {
+    var sdk = cli_util.getSdkDir();
+
+    switch (p.url.basename(id.path)) {
+      case 'dart_sdk.js':
+        var sdkAmdJsPath =
+            p.url.join(sdk.path, 'lib/dev_compiler/amd/dart_sdk.js');
+        return new Asset.fromFile(id, new File(sdkAmdJsPath));
+      case 'require.js':
+        var requireJsPath =
+            p.url.join(sdk.path, 'lib/dev_compiler/amd/require.js');
+        return new Asset.fromFile(id, new File(requireJsPath));
+      default:
+        return null;
+    }
+  }
+
+  /// Whether or not this looks like a request for an entrypoint or bootstrap
+  /// file.
+  bool _isEntrypointId(AssetId id) =>
+      id.path.endsWith('.bootstrap.js') ||
+      id.path.endsWith('.bootstrap.js.map') ||
+      id.path.endsWith('.dart.js') ||
+      id.path.endsWith('.dart.js.map');
+
+  /// Helper to read a module config file, used by [_moduleReader].
+  ///
+  /// Skips barback and reads directly from [this] since we create all these
+  /// files.
+  Future<String> _readModule(AssetId moduleConfigId) async {
+    var asset = await getAssetById(moduleConfigId);
+    return asset.readAsString();
+  }
+
+  /// Gets an [Asset] by [id] asynchronously.
+  ///
+  /// All `.dart` files are read from [_barback], and all other files are read
+  /// from [this].
```

Comment:
I assume this is because we know this will never be called for any file that could be produced by barback aside from .dart ones? Might be good to call that out.